File: libs/pipeline-expression/src/node_check.rs
//! Per-node validation entry point used by `pipeline-authoring-service`
//! to power `POST /api/v1/pipelines/{id}/validate`.
//!
//! The function accepts the pipeline DAG as a `serde_json::Value` (an
//! array of node objects shaped like the persisted JSON) so the lib
//! does not need to know the service's full `PipelineNode` Rust type.
//! It returns a [`PipelineValidationReport`] with one entry per node.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::catalog::transform_signature;
use crate::infer::{ColumnEnv, TypeError, infer_expr};
use crate::parser::parse_expr;
use crate::types::PipelineType;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct NodeValidationError {
    pub node_id: String,
    pub column: Option<String>,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct NodeValidationReport {
    pub node_id: String,
    pub status: String,
    pub errors: Vec<NodeValidationError>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct PipelineValidationReport {
    pub pipeline_id: String,
    pub all_valid: bool,
    pub nodes: Vec<NodeValidationReport>,
}

const STATUS_VALID: &str = "VALID";
const STATUS_INVALID: &str = "INVALID";

/// Validate every node in `nodes_json` (expected to be a JSON array of
/// node objects with `id`, `transform_type`, `config`, `depends_on`
/// fields). Returns one report per node — the order matches the input.
pub fn validate_nodes_json(pipeline_id: &str, nodes_json: &Value) -> PipelineValidationReport {
    let empty: Vec<Value> = Vec::new();
    let nodes = nodes_json.as_array().unwrap_or(&empty);

    let mut node_reports = Vec::with_capacity(nodes.len());
    for node in nodes {
        node_reports.push(validate_one(node, nodes));
    }
    let all_valid = node_reports.iter().all(|r| r.status == STATUS_VALID);
    PipelineValidationReport {
        pipeline_id: pipeline_id.to_string(),
        all_valid,
        nodes: node_reports,
    }
}

fn validate_one(node: &Value, all_nodes: &[Value]) -> NodeValidationReport {
    let id = node_id(node).unwrap_or_default();
    let transform = node
        .get("transform_type")
        .and_then(Value::as_str)
        .unwrap_or("");
    let config = node.get("config").cloned().unwrap_or(Value::Null);
    let depends_on: Vec<String> = node
        .get("depends_on")
        .and_then(Value::as_array)
        .map(|a| a.iter().filter_map(|v| v.as_str().map(String::from)).collect())
        .unwrap_or_default();

    let env = synth_env(&depends_on, all_nodes);
    let mut errors = Vec::new();

    match transform.to_ascii_lowercase().as_str() {
        "passthrough" => {}
        "filter" => check_filter(&id, &config, &env, &mut errors),
        "cast" | "title_case" | "clean_string" => {
            check_columns_array(&id, transform, &config, &mut errors);
        }
        "join" => {
            check_required_string(&id, "how", &config, &mut errors);
            check_required_array(&id, "on", &config, &mut errors);
        }
        "union" => {
            if depends_on.len() < 2 {
                errors.push(NodeValidationError {
                    node_id: id.clone(),
                    column: None,
                    message: "union requires at least 2 upstream nodes".into(),
                });
            }
        }
        "group_by" => {
            check_required_string_array(&id, "keys", &config, &mut errors);
            check_required_array(&id, "aggregations", &config, &mut errors);
        }
        "window" => {
            check_required_string_array(&id, "partition_by", &config, &mut errors);
            check_required_string_array(&id, "order_by", &config, &mut errors);
        }
        "pivot" => {
            check_required_string(&id, "pivot_column", &config, &mut errors);
            check_required_string(&id, "value_column", &config, &mut errors);
        }
        other => {
            // Code transforms (sql/python/llm/wasm) and media transforms
            // are checked elsewhere — only verify the catalog has heard
            // of the name.
            if transform_signature(other).is_none() && !is_known_other(other) {
                // Unknown transform type — surface a single error.
                errors.push(NodeValidationError {
                    node_id: id.clone(),
                    column: None,
                    message: format!("unknown transform_type '{transform}'"),
                });
            }
        }
    }

    let status = if errors.is_empty() {
        STATUS_VALID
    } else {
        STATUS_INVALID
    };
    NodeValidationReport {
        node_id: id,
        status: status.to_string(),
        errors,
    }
}

fn is_known_other(name: &str) -> bool {
    matches!(
        name,
        "sql"
            | "python"
            | "llm"
            | "wasm"
            | "media_set_input"
            | "media_set_output"
            | "media_transform"
            | "convert_media_set_to_table_rows"
            | "get_media_references"
            | "SYNC"
            | "HEALTH_CHECK"
            | "ANALYTICAL"
            | "EXPORT"
    )
}

fn check_filter(
    node_id: &str,
    config: &Value,
    env: &ColumnEnv,
    errors: &mut Vec<NodeValidationError>,
) {
    let predicate = match config.get("predicate").and_then(Value::as_str) {
        Some(s) if !s.trim().is_empty() => s.to_string(),
        _ => {
            errors.push(NodeValidationError {
                node_id: node_id.to_string(),
                column: None,
                message: "filter requires a non-empty `predicate` string".into(),
            });
            return;
        }
    };

    let parsed = match parse_expr(&predicate) {
        Ok(p) => p,
        Err(e) => {
            errors.push(NodeValidationError {
                node_id: node_id.to_string(),
                column: None,
                message: format!("predicate parse error: {e}"),
            });
            return;
        }
    };

    // Empty env => the upstream nodes don't expose schema info yet;
    // suppress UnknownColumn errors so the squiggle UI doesn't spam.
    let permissive = env.is_empty();

    match infer_expr(&parsed, env) {
        Ok(t) => {
            if !matches!(t, PipelineType::Boolean) {
                errors.push(NodeValidationError {
                    node_id: node_id.to_string(),
                    column: None,
                    message: format!("predicate must return Boolean, got {t:?}"),
                });
            }
        }
        Err(type_errors) => {
            for type_error in type_errors {
                if permissive && matches!(type_error, TypeError::UnknownColumn(_)) {
                    continue;
                }
                errors.push(NodeValidationError {
                    node_id: node_id.to_string(),
                    column: column_from_type_error(&type_error),
                    message: type_error.to_string(),
                });
            }
        }
    }
}

fn column_from_type_error(error: &TypeError) -> Option<String> {
    match error {
        TypeError::UnknownColumn(name) => Some(name.clone()),
        _ => None,
    }
}

fn check_columns_array(
    node_id: &str,
    transform: &str,
    config: &Value,
    errors: &mut Vec<NodeValidationError>,
) {
    match config.get("columns") {
        Some(Value::Array(arr)) if arr.iter().all(Value::is_string) && !arr.is_empty() => {}
        Some(_) => errors.push(NodeValidationError {
            node_id: node_id.to_string(),
            column: None,
            message: format!(
                "{transform} requires `columns` to be a non-empty array of strings"
            ),
        }),
        None => errors.push(NodeValidationError {
            node_id: node_id.to_string(),
            column: None,
            message: format!("{transform} requires a `columns` config key"),
        }),
    }
}

fn check_required_string_array(
    node_id: &str,
    key: &str,
    config: &Value,
    errors: &mut Vec<NodeValidationError>,
) {
    match config.get(key) {
        Some(Value::Array(arr)) if arr.iter().all(Value::is_string) => {}
        _ => errors.push(NodeValidationError {
            node_id: node_id.to_string(),
            column: None,
            message: format!("`{key}` must be an array of strings"),
        }),
    }
}

fn check_required_array(
    node_id: &str,
    key: &str,
    config: &Value,
    errors: &mut Vec<NodeValidationError>,
) {
    match config.get(key) {
        Some(Value::Array(_)) => {}
        _ => errors.push(NodeValidationError {
            node_id: node_id.to_string(),
            column: None,
            message: format!("`{key}` must be an array"),
        }),
    }
}

fn check_required_string(
    node_id: &str,
    key: &str,
    config: &Value,
    errors: &mut Vec<NodeValidationError>,
) {
    match config.get(key) {
        Some(Value::String(s)) if !s.trim().is_empty() => {}
        _ => errors.push(NodeValidationError {
            node_id: node_id.to_string(),
            column: None,
            message: format!("`{key}` must be a non-empty string"),
        }),
    }
}

fn synth_env(depends_on: &[String], all_nodes: &[Value]) -> ColumnEnv {
    let mut env = ColumnEnv::new();
    for upstream_id in depends_on {
        let Some(upstream) = all_nodes.iter().find(|n| node_id(n).as_deref() == Some(upstream_id))
        else {
            continue;
        };
        let cfg = upstream.get("config").cloned().unwrap_or(Value::Null);
        if let Some(arr) = cfg.get("columns").and_then(Value::as_array) {
            for col in arr {
                if let Some(name) = col.as_str() {
                    env.insert(name, PipelineType::String);
                }
            }
        }
        if let Some(arr) = cfg.get("output_columns").and_then(Value::as_array) {
            for col in arr {
                if let Some(name) = col.as_str() {
                    env.insert(name, PipelineType::String);
                }
            }
        }
    }
    env
}

fn node_id(node: &Value) -> Option<String> {
    node.get("id")
        .and_then(Value::as_str)
        .map(str::to_string)
}

File: openfoundry-go/libs/pipeline-expression/node_check.go
package pipelineexpression

import (
	"encoding/json"
	"fmt"
	"strings"
)

// NodeValidationError is one validation failure attached to a single
// pipeline node.
type NodeValidationError struct {
	NodeID  string  `json:"node_id"`
	Column  *string `json:"column"`
	Message string  `json:"message"`
}

// NodeValidationReport is the per-node result of [ValidateNodesJSON].
type NodeValidationReport struct {
	NodeID string                `json:"node_id"`
	Status string                `json:"status"`
	Errors []NodeValidationError `json:"errors"`
}

// PipelineValidationReport is the full pipeline result of
// [ValidateNodesJSON].
type PipelineValidationReport struct {
	PipelineID string                 `json:"pipeline_id"`
	AllValid   bool                   `json:"all_valid"`
	Nodes      []NodeValidationReport `json:"nodes"`
}

const (
	statusValid   = "VALID"
	statusInvalid = "INVALID"
)

// ValidateNodesJSON walks the pipeline DAG (an array of node objects
// shaped like the persisted JSON) and returns one report per node, in
// input order.
//
// Mirrors the Rust public surface so `pipeline-authoring-service` can
// call it from the Go port without changes to the wire format.
func ValidateNodesJSON(pipelineID string, nodesJSON json.RawMessage) PipelineValidationReport {
	nodes := parseNodesArray(nodesJSON)

	nodeReports := make([]NodeValidationReport, 0, len(nodes))
	for _, node := range nodes {
		nodeReports = append(nodeReports, validateOne(node, nodes))
	}
	allValid := true
	for _, r := range nodeReports {
		if r.Status != statusValid {
			allValid = false
			break
		}
	}
	return PipelineValidationReport{
		PipelineID: pipelineID,
		AllValid:   allValid,
		Nodes:      nodeReports,
	}
}

// parseNodesArray decodes the raw nodes array. Anything that doesn't
// decode as a JSON array becomes an empty slice — matches the Rust
// `nodes_json.as_array().unwrap_or(&empty)` behaviour.
func parseNodesArray(raw json.RawMessage) []json.RawMessage {
	if len(raw) == 0 {
		return nil
	}
	var arr []json.RawMessage
	if err := json.Unmarshal(raw, &arr); err != nil {
		return nil
	}
	return arr
}

func validateOne(node json.RawMessage, allNodes []json.RawMessage) NodeValidationReport {
	id := nodeID(node)
	transform := stringField(node, "transform_type")
	configRaw := rawField(node, "config")
	dependsOn := stringArrayField(node, "depends_on")

	env := synthEnv(dependsOn, allNodes)
	var errors []NodeValidationError

	switch strings.ToLower(transform) {
	case "passthrough":
		// no-op
	case "filter":
		checkFilter(id, configRaw, env, &errors)
	case "cast", "title_case", "clean_string":
		checkColumnsArray(id, transform, configRaw, &errors)
	case "join":
		checkRequiredString(id, "how", configRaw, &errors)
		checkRequiredArray(id, "on", configRaw, &errors)
	case "union":
		if len(dependsOn) < 2 {
			errors = append(errors, NodeValidationError{
				NodeID:  id,
				Message: "union requires at least 2 upstream nodes",
			})
		}
	case "group_by":
		checkRequiredStringArray(id, "keys", configRaw, &errors)
		checkRequiredArray(id, "aggregations", configRaw, &errors)
	case "window":
		checkRequiredStringArray(id, "partition_by", configRaw, &errors)
		checkRequiredStringArray(id, "order_by", configRaw, &errors)
	case "pivot":
		checkRequiredString(id, "pivot_column", configRaw, &errors)
		checkRequiredString(id, "value_column", configRaw, &errors)
	default:
		other := strings.ToLower(transform)
		if _, ok := TransformSignatureFor(other); !ok && !isKnownOther(other) {
			errors = append(errors, NodeValidationError{
				NodeID:  id,
				Message: fmt.Sprintf("unknown transform_type '%s'", transform),
			})
		}
	}

	status := statusValid
	if len(errors) > 0 {
		status = statusInvalid
	}
	if errors == nil {
		errors = []NodeValidationError{}
	}
	return NodeValidationReport{
		NodeID: id,
		Status: status,
		Errors: errors,
	}
}

func isKnownOther(name string) bool {
	switch name {
	case "sql", "python", "llm", "wasm",
		"media_set_input", "media_set_output", "media_transform",
		"convert_media_set_to_table_rows", "get_media_references",
		"sync", "health_check", "analytical", "export":
		return true
	}
	return false
}

func checkFilter(nodeID string, config json.RawMessage, env ColumnEnv, errors *[]NodeValidationError) {
	predicate := stringField(config, "predicate")
	if strings.TrimSpace(predicate) == "" {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: "filter requires a non-empty `predicate` string",
		})
		return
	}

	parsed, err := ParseExpr(predicate)
	if err != nil {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("predicate parse error: %s", err.Error()),
		})
		return
	}

	// Empty env => the upstream nodes don't expose schema info yet;
	// suppress UnknownColumn errors so the squiggle UI doesn't spam.
	permissive := env.IsEmpty()

	t, typeErrors := InferExpr(parsed, env)
	if len(typeErrors) == 0 {
		if t.Kind != KindBoolean {
			*errors = append(*errors, NodeValidationError{
				NodeID:  nodeID,
				Message: fmt.Sprintf("predicate must return Boolean, got %s", typeHintDebug(t)),
			})
		}
		return
	}
	for _, te := range typeErrors {
		if permissive && te.Kind == TypeErrUnknownColumn {
			continue
		}
		col := columnFromTypeError(te)
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Column:  col,
			Message: te.Error(),
		})
	}
}

func columnFromTypeError(err TypeError) *string {
	if err.Kind == TypeErrUnknownColumn {
		s := err.Detail
		return &s
	}
	return nil
}

func checkColumnsArray(nodeID, transform string, config json.RawMessage, errors *[]NodeValidationError) {
	raw := rawField(config, "columns")
	if len(raw) == 0 || string(raw) == "null" {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("%s requires a `columns` config key", transform),
		})
		return
	}
	var arr []json.RawMessage
	if err := json.Unmarshal(raw, &arr); err != nil {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("%s requires `columns` to be a non-empty array of strings", transform),
		})
		return
	}
	if len(arr) == 0 {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("%s requires `columns` to be a non-empty array of strings", transform),
		})
		return
	}
	for _, raw := range arr {
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			*errors = append(*errors, NodeValidationError{
				NodeID:  nodeID,
				Message: fmt.Sprintf("%s requires `columns` to be a non-empty array of strings", transform),
			})
			return
		}
	}
}

func checkRequiredStringArray(nodeID, key string, config json.RawMessage, errors *[]NodeValidationError) {
	raw := rawField(config, key)
	if len(raw) == 0 {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("`%s` must be an array of strings", key),
		})
		return
	}
	var arr []json.RawMessage
	if err := json.Unmarshal(raw, &arr); err != nil {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("`%s` must be an array of strings", key),
		})
		return
	}
	for _, item := range arr {
		var s string
		if err := json.Unmarshal(item, &s); err != nil {
			*errors = append(*errors, NodeValidationError{
				NodeID:  nodeID,
				Message: fmt.Sprintf("`%s` must be an array of strings", key),
			})
			return
		}
	}
}

func checkRequiredArray(nodeID, key string, config json.RawMessage, errors *[]NodeValidationError) {
	raw := rawField(config, key)
	if len(raw) == 0 {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("`%s` must be an array", key),
		})
		return
	}
	var arr []json.RawMessage
	if err := json.Unmarshal(raw, &arr); err != nil {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("`%s` must be an array", key),
		})
	}
}

func checkRequiredString(nodeID, key string, config json.RawMessage, errors *[]NodeValidationError) {
	v := stringField(config, key)
	if strings.TrimSpace(v) == "" {
		*errors = append(*errors, NodeValidationError{
			NodeID:  nodeID,
			Message: fmt.Sprintf("`%s` must be a non-empty string", key),
		})
	}
}

func synthEnv(dependsOn []string, allNodes []json.RawMessage) ColumnEnv {
	env := NewColumnEnv()
	for _, upstreamID := range dependsOn {
		var upstream json.RawMessage
		for _, n := range allNodes {
			if nodeID(n) == upstreamID {
				upstream = n
				break
			}
		}
		if len(upstream) == 0 {
			continue
		}
		cfg := rawField(upstream, "config")
		for _, key := range []string{"columns", "output_columns"} {
			raw := rawField(cfg, key)
			if len(raw) == 0 {
				continue
			}
			var arr []json.RawMessage
			if err := json.Unmarshal(raw, &arr); err != nil {
				continue
			}
			for _, item := range arr {
				var s string
				if err := json.Unmarshal(item, &s); err != nil {
					continue
				}
				env.Insert(s, StringType())
			}
		}
	}
	return env
}

// nodeID extracts the `id` field from a JSON node object. Returns ""
// when the field is missing or not a string.
func nodeID(node json.RawMessage) string {
	return stringField(node, "id")
}

func stringField(raw json.RawMessage, key string) string {
	if len(raw) == 0 {
		return ""
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return ""
	}
	field, ok := obj[key]
	if !ok {
		return ""
	}
	var s string
	if err := json.Unmarshal(field, &s); err != nil {
		return ""
	}
	return s
}

func stringArrayField(raw json.RawMessage, key string) []string {
	if len(raw) == 0 {
		return nil
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil
	}
	field, ok := obj[key]
	if !ok {
		return nil
	}
	var arr []json.RawMessage
	if err := json.Unmarshal(field, &arr); err != nil {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, item := range arr {
		var s string
		if err := json.Unmarshal(item, &s); err != nil {
			continue
		}
		out = append(out, s)
	}
	return out
}

func rawField(raw json.RawMessage, key string) json.RawMessage {
	if len(raw) == 0 {
		return nil
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return nil
	}
	return obj[key]
}
